### PR TITLE
Save Computation in Known Leaf Nodes, Fix minChildWeight bug

### PR DESCRIPTION
**Pull request by samanthacampo**
### Description
Save Computation in Known Leaf Nodes - makes leaf nodes immediately instead of creating TrainingNodes first and then converting them to LeafNodes. 
Fix minChildWeight bug - minChildWeight was previously being compared to the unweighted number of examples, but will now correctly be compared to the sum of weighted examples.

### Motivation
Save computation. Fix a bug.

### CODE REVIEW

File: Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java [hunk 1]
```diff
@@ -41,33 +43,71 @@
...
-     *                                  minImpurityDecrease.
      * @return A possibly empty list of TrainingNodes.
      */
     public abstract List<AbstractTrainingNode<T>> buildTree(int[] featureIDs, SplittableRandom rng,
-                                                            boolean useRandomSplitPoints, float scaledMinImpurityDecrease);
+                                                            boolean useRandomSplitPoints);
 
     public abstract Node<T> convertTree();
 
+    public abstract float getWeightSum();
```
Comment (Craigacp) [inline Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java:68]: Bit of javadoc for the new method please.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java [hunk 2]
```diff
@@ -272,11 +314,28 @@ private RegressorTrainingNode(RegressorImpurity impurity, ArrayList<TreeFeature>
...
-        greaterThan = new RegressorTrainingNode(impurity, greaterThanData, rightIndices, targets, weights, dimName, rightIndices.length, depth + 1, featureIDMap, labelIDMap);
+
         List<AbstractTrainingNode<Regressor>> output = new ArrayList<>(2);
-        output.add(lessThanOrEqual);
-        output.add(greaterThan);
+        AbstractTrainingNode<Regressor> tmpNode;
+        if (shouldMakeLeftLeaf) {
+            lessThanOrEqual = mkLeaf(leftImpurityScore, leftIndices);
+        }
+        else {
```
Comment (Craigacp) [inline Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:323]: else formatting

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java [hunk 3]
```diff
@@ -272,11 +314,28 @@ private RegressorTrainingNode(RegressorImpurity impurity, ArrayList<TreeFeature>
...
+            tmpNode = new RegressorTrainingNode(impurity, lessThanData, leftIndices, targets, weights, dimName,
+                    leftIndices.length, depth + 1, featureIDMap, labelIDMap, leafDeterminer, leftWeightSum, leftImpurityScore);
+            lessThanOrEqual = tmpNode;
+            output.add(tmpNode);
+        }
+
+        if (shouldMakeRightLeaf) {
+            greaterThan = mkLeaf(rightImpurityScore, rightIndices);
+        }
+        else {
```
Comment (Craigacp) [inline Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:333]: else formatting

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java [hunk 4]
```diff
@@ -287,28 +346,34 @@ private RegressorTrainingNode(RegressorImpurity impurity, ArrayList<TreeFeature>
...
-                double oldMean = mean;
-                mean += (weight / weightSum) * (value - oldMean);
-                variance += weight * (value - oldMean) * (value - mean);
-            }
-            variance = indices.length > 1 ? variance / (weightSum-1) : 0;
-            DimensionTuple leafPred = new DimensionTuple(dimName,mean,variance);
-            return new LeafNode<>(getImpurity(),leafPred,Collections.emptyMap(),false);
+            return mkSplitNode();
+        }
+        return mkLeaf(getImpurity(), indices);
```
Comment (Craigacp) [inline Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:351]: else formatting

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java [hunk 5]
```diff
@@ -287,28 +346,34 @@ private RegressorTrainingNode(RegressorImpurity impurity, ArrayList<TreeFeature>
...
+    }
+
+    /**
+     * Makes a {@link LeafNode}
+     * @param impurityScore the impurity score for the node.
+     * @param indices the indices of the examples to be placed in the node.
+     * @return A {@link LeafNode}
+     */
+    private LeafNode<Regressor> mkLeaf(double impurityScore, int[] indices) {
```
Comment (Craigacp) [inline Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:360]: Shadows `indices` field.

**Review by Craigacp — CHANGES_REQUESTED** [review]
A bunch of minor things, mostly renaming, formatting and a bit of javadoc.

File: Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java [hunk 6]
```diff
@@ -67,58 +67,86 @@
 
     private final float[] weightedLabelCounts;
 
+    private final float weightSum;
+
     /**
      * Constructor which creates the inverted file.
      * @param impurity The impurity function to use.
      * @param examples The training data.
+     * @param leafDeterminer Contains parameters needed to determine whether a node is a leaf.
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java:76]: Done.

File: Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java [hunk 7]
```diff
@@ -254,47 +304,63 @@ private ClassifierTrainingNode(LabelImpurity impurity, ArrayList<TreeFeature> da
...
 
-        lessThanOrEqual = new ClassifierTrainingNode(impurity, lessThanData, lessThanIndices.size, depth + 1, featureIDMap, labelIDMap);
-        greaterThan = new ClassifierTrainingNode(impurity, greaterThanData, numExamples - lessThanIndices.size, depth + 1, featureIDMap, labelIDMap);
         List<AbstractTrainingNode<Label>> output = new ArrayList<>(2);
-        output.add(lessThanOrEqual);
-        output.add(greaterThan);
+        AbstractTrainingNode<Label> tmpNode;
+        if (shouldMakeLessThanLeaf) {
+            lessThanOrEqual = mkLeaf(lessThanImpurityScore, lessThanCounts);
+        }
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java:311]: Done.

File: Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java [hunk 8]
```diff
@@ -254,47 +304,63 @@ private ClassifierTrainingNode(LabelImpurity impurity, ArrayList<TreeFeature> da
...
+        else {
+            tmpNode = new ClassifierTrainingNode(impurity, lessThanData, lessThanIndices.size, depth + 1,
+                    featureIDMap, labelIDMap, leafDeterminer, lessThanCounts, lessThanWeightSum, lessThanImpurityScore);
+            lessThanOrEqual = tmpNode;
+            output.add(tmpNode);
+        }
+
+        if (shouldMakeGreaterThanLeaf) {
+            greaterThan = mkLeaf(greaterThanImpurityScore, greaterThanCounts);
+        }
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java:321]: Done.

File: Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java [hunk 9]
```diff
@@ -254,47 +304,63 @@ private ClassifierTrainingNode(LabelImpurity impurity, ArrayList<TreeFeature> da
...
-            }
-            return new LeafNode<>(impurity.impurity(weightedLabelCounts),maxLabel,counts,true);
+            return mkSplitNode();
         }
-    }
-
-    @Override
-    public double getImpurity() {
-        return impurityScore;
+        return mkLeaf(getImpurity(), weightedLabelCounts);
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Classification/DecisionTree/src/main/java/org/tribuo/classification/dtree/impl/ClassifierTrainingNode.java:363]: Done.

File: Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java [hunk 10]
```diff
@@ -41,33 +43,71 @@
...
+
     public int getDepth() {
         return depth;
     }
 
+    public int getNumExamples() {
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java:74]: Done.

File: Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java (hunk 1 above)
Comment (samanthacampo) [reply Common/Trees/src/main/java/org/tribuo/common/tree/AbstractTrainingNode.java:68]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java (hunk 2 above)
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:323]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java (hunk 3 above)
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:333]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java (hunk 4 above)
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:351]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java (hunk 5 above)
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/RegressorTrainingNode.java:360]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 11]
```diff
@@ -309,13 +351,30 @@ private double calcImpurity() {
...
-                weights, rightIndices.length, depth + 1, featureIDMap, labelIDMap, normalize);
+
         List<AbstractTrainingNode<Regressor>> output = new ArrayList<>(2);
-        output.add(lessThanOrEqual);
-        output.add(greaterThan);
+        AbstractTrainingNode<Regressor> tmpNode;
+        if (shouldMakeLeftLeaf) {
+            lessThanOrEqual = mkLeaf(leftImpurityScore, leftIndices);
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:360]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 12]
```diff
@@ -309,13 +351,30 @@ private double calcImpurity() {
...
+                    weights, leftIndices.length, depth + 1, featureIDMap, labelIDMap, normalize, leafDeterminer,
+                    leftWeightSum, leftImpurityScore);
+            lessThanOrEqual = tmpNode;
+            output.add(tmpNode);
+        }
+
+        if (shouldMakeRightLeaf) {
+            greaterThan = mkLeaf(rightImpurityScore, rightIndices);
+        }
+        else {
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:371]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 13]
```diff
@@ -326,60 +385,66 @@ private double calcImpurity() {
...
-                    for (int j = 0; j < targets.length; j++) {
-                        float value = targets[j][idx];
-
-                        double oldMean = mean[j];
-                        mean[j] += (weight / weightSum) * (value - oldMean);
-                        variance[j] += weight * (value - oldMean) * (value - mean[j]);
-                    }
+            return mkSplitNode();
+        }
+        return mkLeaf(getImpurity(), indices);
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:390]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 14]
```diff
@@ -326,60 +385,66 @@ private double calcImpurity() {
...
+    }
+
+    /**
+     * Makes a {@link LeafNode}
+     * @param impurityScore the impurity score for the node.
+     * @param indices the indices of the examples to be placed in the node.
+     * @return A {@link LeafNode}
+     */
+    private LeafNode<Regressor> mkLeaf(double impurityScore, int[] indices) {
+        double weightSum = 0.0;
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:400]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 15]
```diff
@@ -97,19 +126,25 @@ private JointRegressorTrainingNode(RegressorImpurity impurity, ArrayList<TreeFea
...
+        return weightSum;
+    }
+
     /**
      * Calculates the impurity score of the node.
-     * @return the impurity score of the node.
+     * @return The impurity score of the node.
      */
-    private double calcImpurity() {
+    private double calcImpurity(int[] indices) {
```
[the remark this answers is not in the document]
Comment (samanthacampo) [reply Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:147]: Done.

File: Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java [hunk 16]
```diff
@@ -144,10 +144,10 @@ public float getWeightSum() {
      * Calculates the impurity score of the node.
      * @return The impurity score of the node.
      */
-    private double calcImpurity(int[] indices) {
+    private double calcImpurity(int[] cur_indices) {
```
Comment (Craigacp) [inline Regression/RegressionTree/src/main/java/org/tribuo/regression/rtree/impl/JointRegressorTrainingNode.java:147]: camelCase not snake_case

**Review by Craigacp — CHANGES_REQUESTED** [review]
Just a variable renaming.

**Review by Craigacp — APPROVED** [review]
LGTM.

**Pull request merged**